streams/src/test/java/org/apache/kafka/streams/integration/KStreamKStreamIntegrationTest.java

```diff
@@ -0,0 +1,186 @@
+        CONSUMER_CONFIG.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+    }
+
+    @AfterAll
+    public static void closeCluster() {
+        CLUSTER.stop();
+    }
+
+    @BeforeEach
+    public void before(final TestInfo testInfo) throws IOException {
+        final String stateDirBasePath = TestUtils.tempDirectory().getPath();
+        final String safeTestName = safeUniqueTestName(getClass(), testInfo);
+        streamsConfig = getStreamsConfig(safeTestName);
+        streamsConfig.put(StreamsConfig.STATE_DIR_CONFIG, stateDirBasePath);
+    }
+
+    @AfterEach
+    public void after() throws IOException {
+        if (streams != null) {
+            streams.close();
+            streams = null;
+        }
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfig);
+    }
+
+    @Test
+    public void shouldOuterJoin() throws Exception {
+        final Set<KeyValue<String, String>> expectedOne = new HashSet<>();
+        expectedOne.add(new KeyValue<>("Key-1", "value1=left-1,value2=null"));
+        expectedOne.add(new KeyValue<>("Key-2", "value1=left-2,value2=null"));
+        expectedOne.add(new KeyValue<>("Key-3", "value1=left-3,value2=null"));
+        expectedOne.add(new KeyValue<>("Key-4", "value1=left-4,value2=null"));
+
+        verifyKStreamKStreamOuterJoin(expectedOne);
+    }
+
+    private void verifyKStreamKStreamOuterJoin(final Set<KeyValue<String, String>> expectedResult) throws Exception {
+        streams = prepareTopology(streamsConfig);
+
+        startApplicationAndWaitUntilRunning(Collections.singletonList(streams), ofSeconds(120));
+
+        PRODUCER_CONFIG.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        PRODUCER_CONFIG.put(ProducerConfig.ACKS_CONFIG, "all");
+        PRODUCER_CONFIG.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        PRODUCER_CONFIG.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+
+        final List<KeyValue<String, String>> stream1 = asList(
+                new KeyValue<>("Key-1", "left-1"),
+                new KeyValue<>("Key-2", "left-2"),
+                new KeyValue<>("Key-3", "left-3"),
+                new KeyValue<>("Key-4", "left-4")
+        );
+
+        final List<KeyValue<String, String>> stream2 = asList(
+                new KeyValue<>("Key-1", "right-1"),
+                new KeyValue<>("Key-2", "right-2"),
+                new KeyValue<>("Key-3", "right-3"),
+                new KeyValue<>("Key-4", "right-4")
+        );
+
+        IntegrationTestUtils.produceKeyValuesSynchronously(STREAM_1, stream1, PRODUCER_CONFIG, MOCK_TIME);
+        MOCK_TIME.sleep(11000);
+        Thread.sleep(11000);
+        IntegrationTestUtils.produceKeyValuesSynchronously(STREAM_1, stream2, PRODUCER_CONFIG, MOCK_TIME);
+
+        final Set<KeyValue<String, String>> result = new HashSet<>(waitUntilMinKeyValueRecordsReceived(
```

Comment:
Looks like the test produces 8 records, but only waits for 4 result records and only validates those first 4? 

Also, how does this test guarantee the order in which the input records are processed, given that there are multiple input partitions? I think I'm missing something :)

Reply:
> Looks like the test produces 8 records, but only waits for 4 result records and only validates those first 4?

Yes, we only get 4 result for the first "batch / window" -- all 4 are outer-join result. The remaining 4 records won't get emitted, because they are "stuck": without new input record, stream-time does not advance, and the join window does not close. -- The goal of the second batch it really just to ensure to flush out the first 4 records.

Reply:
> Also, how does this test guarantee the order in which the input records are processed, given that there are multiple input partitions? I think I'm missing something :)

We cannot. That is why a `Set` is used to compare expected to retrieved results.

Reply:
Ah! I missed that. Thanks for clearing it up.